File: ruby/lib/jugador_especulador.rb
require_relative ('diario')
module Civitas
    class JugadorEspeculador < Jugador
        @FactorEspeculador=2
        @CasasMax=@FactorEspeculador*Jugador.CasasMax
        @HotelesMax=@FactorEspeculador*Jugador.HotelesMax

        def initialize(fianza)
            super("nuevo")
            @fianza=fianza  
        end

        def self.nuevoEspeculador(jugador, fianza)
            Diario.instance.ocurreEvento("Convertido a JUGADOR ESPECULADOR")
            ret=new(fianza)
            ret.copia(jugador)
            for i in jugador.propiedades do
                i.actualizaPropietarioPorConversion(ret)
            end
            return ret
        end
        private_class_method :new

        def debeSerEncarcelado
            if @encarcelado
                return false
            elsif tieneSalvoConducto
                perderSalvoConducto
                Diario.instance.ocurreEvento("El jugador "+@nombre+" se libra de la cárcel")
                return false
            elsif puedePagarFianza
                paga(@fianza)
                return false
            end

            return true
        end

        def puedePagarFianza
            return @saldo >= @fianza
        end

        def toString
            str=super
            str+="JUGADOR ESPECULADOR"
        end

        def pagaImpuesto(cantidad)
            if @encarcelado 
                return false 
            end
            return paga(cantidad/2)
        end

    end
end

File: java/src/civitas/JugadorEspeculador.java
package civitas;

public class JugadorEspeculador extends Jugador {
    private static int factorEspeculador;
    private int fianza;

    JugadorEspeculador(Jugador jugador, int fianza) {
        super(jugador);
        factorEspeculador = 2;
        this.fianza = fianza;

        for (int i = 0; i < propiedades.size(); i++) {
            propiedades.get(i).actualizaPropietarioPorConversion(this);
        }
    }

    @Override
    protected int getCasasMax() {
        return CasasMax * factorEspeculador;
    }

    @Override
    protected int getHotelesMax() {
        return HotelesMax * factorEspeculador;
    }

    @Override
    protected boolean debeSerEncarcelado() {
        if (encarcelado) {
            return false;
        }

        else if (tieneSalvoconducto()) {
            perderSalvoconducto();
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " ha perdido su salvoconducto");
            return false;
        }

        else {
            if (puedoGastar(fianza)) {
                paga(fianza);
                return false;
            }
            return true;
        }
    }

    @Override
    boolean pagaImpuesto(float cantidad) {
        if (encarcelado)
            return false;
        return paga(cantidad / 2);
    }

    @Override
    public String toString() {
        String props = "";
        for (int i = 0; i < propiedades.size(); i++) {
            props += propiedades.get(i).getNombre() + " ";
        }
        String s = "Nombre: " + nombre + "Es un jugador especulador " + " Saldo: " + saldo + " Propiedades: " + props + " Casilla actual: "
                + numCasillaActual;
        return s;
    }


}